pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1277,74 +1279,87 @@ func getMatchingTopologyPairs(pod *v1.Pod, nodeInfoMap map[string]*schedulercach
 			catchError(fmt.Errorf("node not found"))
 			return
 		}
-		nodeTopologyMaps := newTopologyPairsMaps()
 		for _, existingPod := range nodeInfo.PodsWithAffinity() {
-			affinity := existingPod.Spec.Affinity
-			if affinity == nil {
-				continue
-			}
-			for _, term := range GetPodAntiAffinityTerms(affinity.PodAntiAffinity) {
-				namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(existingPod, &term)
-				selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
-				if err != nil {
-					catchError(err)
-					return
-				}
-				if priorityutil.PodMatchesTermsNamespaceAndSelector(pod, namespaces, selector) {
-					if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
-						pair := topologyPair{key: term.TopologyKey, value: topologyValue}
-						nodeTopologyMaps.addTopologyPair(pair, existingPod)
-					}
-				}
+			existingPodTopologyMaps, err := getMatchingTopologyPairsOfExistingPod(pod, existingPod, node)
+			if err != nil {
+				catchError(err)
+				return
 			}
-			if len(nodeTopologyMaps.podToTopologyPairs) > 0 {
-				appendTopologyPairsMaps(nodeTopologyMaps)
+
+			if len(existingPodTopologyMaps.podToTopologyPairs) > 0 {
+				appendTopologyPairsMaps(existingPodTopologyMaps)
 			}
 		}
 	}
 	workqueue.Parallelize(16, len(allNodeNames), processNode)
 	return topologyMaps, firstError
 }
 
+// getMatchingTopologyPairsOfExistingPod is a helper to calculate on each existing pod
+// of given node, whether (1) it has PodAntiAffinity and (2) AffinityTerms matches incoming pod
 func getMatchingTopologyPairsOfExistingPod(newPod *v1.Pod, existingPod *v1.Pod, node *v1.Node) (*topologyPairsMaps, error) {
 	topologyMaps := newTopologyPairsMaps()
 	affinity := existingPod.Spec.Affinity
-	if affinity != nil && affinity.PodAntiAffinity != nil {
-		for _, term := range GetPodAntiAffinityTerms(affinity.PodAntiAffinity) {
-			namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(existingPod, &term)
-			selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
-			if err != nil {
-				return nil, err
-			}
-			if priorityutil.PodMatchesTermsNamespaceAndSelector(newPod, namespaces, selector) {
-				if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
-					pair := topologyPair{key: term.TopologyKey, value: topologyValue}
-					topologyMaps.addTopologyPair(pair, existingPod)
-				}
-			}
+	if affinity == nil || affinity.PodAntiAffinity == nil {
+		return topologyMaps, nil
+	}
+
+	matchMap := make(map[string]bool) // key is TopologyKey
+	terms := GetPodAntiAffinityTerms(affinity.PodAntiAffinity)
+	for _, term := range terms {
+		// if topologyKey in term is invalid, simply return as terms of existingPod
+		// have no chance to match labels of incomingPod
+		if _, ok := node.Labels[term.TopologyKey]; !ok {
```

Comment:
This is not true. If a term topologyKey does not exist on the node, we should still check that the pod matches the term and other terms. This term does not add any pair to our topology pairs though.

Reply:
No, the term is not bypassed. Although our API is not very clear, but if you think about the actual use-case, when a user puts anti-affinity and provides some terms on the anti-affinity rule, the user expects other pods to match all the rules. The user does not know at the time of pod creation which node the pod will land on, but they still expect other pods to match all the terms, although the node may not have the label specified on the term. So, we should not bypass the term and should still check the term selectors and if the selectors don't match the pod, the anti-affinity/affinity rule does not apply to the pod.